Aria scan fails with "filter must be present if outputRequired is false"

An error occurs if filter pushdown is enabled and filter expression can be optimized in a way that reduces the set of inputs: 

```
> set session hive.pushdown_filter_enabled=true;
SET SESSION

> SELECT partkey FROM lineitem WHERE orderkey > 10 OR if(json_extract(json_parse('{}'), '$.a') IS NOT NULL, quantity * discount) > 0;

Query 20200121_154338_00012_v6fkz failed: Error opening Hive split file:/Users/mbasmanova/aria/test_data/hive_data/tpch/lineitem/20190723_181544_00002_ea4mu_628f36fa-c8f1-472d-9f9c-e8be512ab8d5 (offset=0, length=140103): filter must be present if outputRequired is false
com.facebook.presto.spi.PrestoException: Error opening Hive split file:/Users/mbasmanova/aria/test_data/hive_data/tpch/lineitem/20190723_181544_00002_ea4mu_628f36fa-c8f1-472d-9f9c-e8be512ab8d5 (offset=0, length=140103): filter must be present if outputRequired is false
	at com.facebook.presto.hive.orc.OrcSelectivePageSourceFactory.createOrcPageSource(OrcSelectivePageSourceFactory.java:407)
	at com.facebook.presto.hive.orc.OrcSelectivePageSourceFactory.createPageSource(OrcSelectivePageSourceFactory.java:223)
	at com.facebook.presto.hive.HivePageSourceProvider.createSelectivePageSource(HivePageSourceProvider.java:220)
	at com.facebook.presto.hive.HivePageSourceProvider.createPageSource(HivePageSourceProvider.java:126)
	at com.facebook.presto.spi.connector.classloader.ClassLoaderSafeConnectorPageSourceProvider.createPageSource(ClassLoaderSafeConnectorPageSourceProvider.java:51)
	at com.facebook.presto.split.PageSourceManager.createPageSource(PageSourceManager.java:58)
	at com.facebook.presto.operator.TableScanOperator.getOutput(TableScanOperator.java:248)
	at com.facebook.presto.operator.Driver.processInternal(Driver.java:379)
	at com.facebook.presto.operator.Driver.lambda$processFor$8(Driver.java:283)
	at com.facebook.presto.operator.Driver.tryWithLock(Driver.java:675)
	at com.facebook.presto.operator.Driver.processFor(Driver.java:276)
	at com.facebook.presto.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:1077)
	at com.facebook.presto.execution.executor.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:162)
	at com.facebook.presto.execution.executor.TaskExecutor$TaskRunner.run(TaskExecutor.java:483)
	at com.facebook.presto.$gen.Presto_null__testversion____20200121_154218_2.run(Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.IllegalArgumentException: filter must be present if outputRequired is false
	at com.google.common.base.Preconditions.checkArgument(Preconditions.java:141)
	at com.facebook.presto.orc.reader.DoubleSelectiveStreamReader.<init>(DoubleSelectiveStreamReader.java:92)
	at com.facebook.presto.orc.reader.SelectiveStreamReaders.createStreamReader(SelectiveStreamReaders.java:71)
	at com.facebook.presto.orc.OrcSelectiveRecordReader.createStreamReaders(OrcSelectiveRecordReader.java:584)
	at com.facebook.presto.orc.OrcSelectiveRecordReader.<init>(OrcSelectiveRecordReader.java:177)
	at com.facebook.presto.orc.OrcReader.createSelectiveRecordReader(OrcReader.java:236)
	at com.facebook.presto.hive.orc.OrcSelectivePageSourceFactory.createOrcPageSource(OrcSelectivePageSourceFactory.java:370)
	... 17 more
```

CC: @prestodb/aria

--- a/presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java
+++ b/presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java
@@ -140,6 +140,7 @@ import java.util.stream.Collectors;
 import java.util.stream.IntStream;
 
 import static com.facebook.airlift.concurrent.MoreFutures.toCompletableFuture;
+import static com.facebook.presto.expressions.LogicalRowExpressions.FALSE_CONSTANT;
 import static com.facebook.presto.expressions.LogicalRowExpressions.TRUE_CONSTANT;
 import static com.facebook.presto.expressions.LogicalRowExpressions.and;
 import static com.facebook.presto.expressions.LogicalRowExpressions.binaryExpression;
@@ -277,6 +278,7 @@ public class HiveMetadata
 
     private static final String PARTITIONS_TABLE_SUFFIX = "$partitions";
     private static final String PRESTO_TEMPORARY_TABLE_NAME_PREFIX = "__presto_temporary_table_";
+    private static final ConnectorTableLayout EMPTY_TABLE_LAYOUT = new ConnectorTableLayout(new ConnectorTableLayoutHandle() {}) {};
 
     private final boolean allowCorruptWritesForTesting;
     private final SemiTransactionalHiveMetastore metastore;
@@ -1874,6 +1876,17 @@ public class HiveMetadata
             decomposedFilter = intersectExtractionResult(decomposedFilter, new ExtractionResult(currentHiveLayout.getDomainPredicate(), currentHiveLayout.getRemainingPredicate()));
         }
 
+        if (decomposedFilter.getTupleDomain().isNone()) {
+            return new ConnectorPushdownFilterResult(EMPTY_TABLE_LAYOUT, FALSE_CONSTANT);
+        }
+
+        RowExpression optimizedRemainingExpression = rowExpressionService.getExpressionOptimizer().optimize(decomposedFilter.getRemainingExpression(), OPTIMIZED, session);
+        if (optimizedRemainingExpression instanceof ConstantExpression) {
+            ConstantExpression constantExpression = (ConstantExpression) optimizedRemainingExpression;
+            if (FALSE_CONSTANT.equals(constantExpression) || constantExpression.getValue() == null) {
+                return new ConnectorPushdownFilterResult(EMPTY_TABLE_LAYOUT, FALSE_CONSTANT);
+            }
+        }
         Map<String, ColumnHandle> columnHandles = getColumnHandles(session, tableHandle);
         TupleDomain<ColumnHandle> entireColumnDomain = decomposedFilter.getTupleDomain()
                 .transform(subfield -> isEntireColumn(subfield) ? subfield.getRootName() : null)
@@ -1911,7 +1924,9 @@ public class HiveMetadata
         domainPredicate.getDomains().get().keySet().stream()
                 .map(Subfield::getRootName)
                 .forEach(predicateColumnNames::add);
-        extractAll(decomposedFilter.getRemainingExpression()).stream()
+        // Include only columns referenced in the optimized expression. Although the expression is sent to the worker node
+        // unoptimized, the worker is expected to optimize the expression before executing.
+        extractAll(optimizedRemainingExpression).stream()
                 .map(VariableReferenceExpression::getName)
                 .forEach(predicateColumnNames::add);
 

--- a/presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java
+++ b/presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java
@@ -85,6 +85,7 @@ import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.node;
 import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.output;
 import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.project;
 import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.strictTableScan;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.values;
 import static com.facebook.presto.sql.planner.optimizations.PlanNodeSearcher.searchFrom;
 import static com.facebook.presto.sql.planner.plan.ExchangeNode.Scope.REMOTE_STREAMING;
 import static com.facebook.presto.sql.planner.plan.ExchangeNode.Type.GATHER;
@@ -181,6 +182,26 @@ public class TestHiveLogicalPlanner
                 output(exchange(project(
                         filter("mod(orderkey, 2) = 1",
                                 strictTableScan("lineitem", identityMap("linenumber", "orderkey")))))));
+        // Remaining predicate is NULL
+        assertPlan(pushdownFilterEnabled, "SELECT linenumber FROM lineitem WHERE cardinality(NULL) > 0",
+                output(values("linenumber")));
+
+        assertPlan(pushdownFilterEnabled, "SELECT linenumber FROM lineitem WHERE orderkey > 10 AND cardinality(NULL) > 0",
+                output(values("linenumber")));
+
+        // Remaining predicate is always FALSE
+        assertPlan(pushdownFilterEnabled, "SELECT linenumber FROM lineitem WHERE cardinality(ARRAY[1]) > 1",
+                output(values("linenumber")));
+
+        assertPlan(pushdownFilterEnabled, "SELECT linenumber FROM lineitem WHERE orderkey > 10 AND cardinality(ARRAY[1]) > 1",
+                output(values("linenumber")));
+
+        // TupleDomain predicate is always FALSE
+        assertPlan(pushdownFilterEnabled, "SELECT linenumber FROM lineitem WHERE orderkey = 1 AND orderkey = 2",
+                output(values("linenumber")));
+
+        assertPlan(pushdownFilterEnabled, "SELECT linenumber FROM lineitem WHERE orderkey = 1 AND orderkey = 2 AND linenumber % 2 = 1",
+                output(values("linenumber")));
 
         FunctionManager functionManager = getQueryRunner().getMetadata().getFunctionManager();
         FunctionResolution functionResolution = new FunctionResolution(functionManager);
@@ -332,6 +353,10 @@ public class TestHiveLogicalPlanner
 
         assertPushdownFilterOnSubfields("SELECT * FROM test_pushdown_filter_on_subfields WHERE c.a IS NOT NULL AND c.c IS NOT NULL",
                 ImmutableMap.of(new Subfield("c.a"), notNull(BIGINT), new Subfield("c.c"), notNull(new ArrayType(BIGINT))));
+
+        // TupleDomain predicate is always FALSE
+        assertPlan(pushdownFilterEnabled(), "SELECT id FROM test_pushdown_filter_on_subfields WHERE c.a = 1 AND c.a = 2",
+                output(values("id")));
 
         assertUpdate("DROP TABLE test_pushdown_filter_on_subfields");
     }

--- a/presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java
+++ b/presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java
@@ -644,6 +644,7 @@ public class TestHivePushdownFilterQueries
         // filter-only struct
         assertQueryUsingH2Cte("SELECT orderkey FROM lineitem_ex WHERE info IS NOT NULL");
         assertQueryUsingH2Cte("SELECT orderkey FROM lineitem_ex WHERE info IS NOT NULL AND info.orderkey = 16515", rewriter);
+        assertQueryReturnsEmptyResult("SELECT orderkey FROM lineitem_ex WHERE info IS NOT NULL AND info.orderkey = 16515 and info.orderkey = 16516");
         assertQueryUsingH2Cte("SELECT orderkey FROM lineitem_ex WHERE info IS NOT NULL AND info.orderkey + 1 = 16514", rewriter);
 
         // filters on subfields
@@ -1031,6 +1032,13 @@ public class TestHivePushdownFilterQueries
         finally {
             assertUpdate("DROP TABLE test_nan");
         }
+    }
+
+    @Test
+    public void testFilterFunctionsWithOptimization()
+    {
+        assertQuery("SELECT partkey FROM lineitem WHERE orderkey > 10 OR if(json_extract(json_parse('{}'), '$.a') IS NOT NULL, quantity * discount) > 0",
+                "SELECT partkey FROM lineitem WHERE orderkey > 10");
     }
 
     private Path getPartitionDirectory(String tableName, String partitionClause)

--- a/presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PickTableLayout.java
+++ b/presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PickTableLayout.java
@@ -69,6 +69,7 @@ import java.util.Set;
 import java.util.function.Function;
 
 import static com.facebook.presto.SystemSessionProperties.isNewOptimizerEnabled;
+import static com.facebook.presto.expressions.LogicalRowExpressions.FALSE_CONSTANT;
 import static com.facebook.presto.expressions.LogicalRowExpressions.TRUE_CONSTANT;
 import static com.facebook.presto.expressions.RowExpressionNodeInliner.replaceExpression;
 import static com.facebook.presto.matching.Capture.newCapture;
@@ -323,17 +324,17 @@ public class PickTableLayout
             expression = predicate;
         }
 
-        // optimize rowExpression and return ValuesNode if false. e.g. 0=1 => false, 1>2 => false, (a = 1 and a = 2) => false
-        DomainTranslator.ExtractionResult<VariableReferenceExpression> translatedExpression = translator.fromPredicate(session.toConnectorSession(), expression, BASIC_COLUMN_EXTRACTOR);
-        if (translatedExpression.getTupleDomain().isNone()) {
-            return new ValuesNode(idAllocator.getNextId(), node.getOutputVariables(), ImmutableList.of());
-        }
-
         BiMap<VariableReferenceExpression, VariableReferenceExpression> symbolToColumnMapping = node.getAssignments().entrySet().stream()
                 .collect(toImmutableBiMap(
                         Map.Entry::getKey,
                         entry -> new VariableReferenceExpression(getColumnName(session, metadata, node.getTable(), entry.getValue()), entry.getKey().getType())));
-        PushdownFilterResult pushdownFilterResult = metadata.pushdownFilter(session, node.getTable(), replaceExpression(expression, symbolToColumnMapping));
+
+        RowExpression replacedExpression = replaceExpression(expression, symbolToColumnMapping);
+        // replaceExpression() may further optimize the expression; if the resulting expression is always false, then return empty Values node
+        if (FALSE_CONSTANT.equals(replacedExpression)) {
+            return new ValuesNode(idAllocator.getNextId(), node.getOutputVariables(), ImmutableList.of());
+        }
+        PushdownFilterResult pushdownFilterResult = metadata.pushdownFilter(session, node.getTable(), replacedExpression);
 
         TableLayout layout = pushdownFilterResult.getLayout();
         if (layout.getPredicate().isNone()) {

